crates/ironrdp-rdpdr/src/pdu/efs.rs

```diff
@@ -0,0 +1,229 @@
+        4 // 2 * u16  = 2 * 2 bytes = 4 bytes
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+#[repr(u16)]
+#[allow(non_camel_case_types)]
+pub enum Component {
+    RDPDR_CTYP_CORE = 0x4472,
+    RDPDR_CTYP_PRN = 0x5052,
+}
+
+impl TryFrom<u16> for Component {
+    type Error = PduError;
+
+    fn try_from(value: u16) -> Result<Self, Self::Error> {
+        match value {
+            0x4472 => Ok(Component::RDPDR_CTYP_CORE),
+            0x5052 => Ok(Component::RDPDR_CTYP_PRN),
+            _ => Err(invalid_message_err!("try_from", "Component", "invalid value")),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+#[repr(u16)]
+#[allow(non_camel_case_types)]
+pub enum PacketId {
+    PAKID_CORE_SERVER_ANNOUNCE = 0x496E,
+    PAKID_CORE_CLIENTID_CONFIRM = 0x4343,
+    PAKID_CORE_CLIENT_NAME = 0x434E,
+    PAKID_CORE_DEVICELIST_ANNOUNCE = 0x4441,
+    PAKID_CORE_DEVICE_REPLY = 0x6472,
+    PAKID_CORE_DEVICE_IOREQUEST = 0x4952,
+    PAKID_CORE_DEVICE_IOCOMPLETION = 0x4943,
+    PAKID_CORE_SERVER_CAPABILITY = 0x5350,
+    PAKID_CORE_CLIENT_CAPABILITY = 0x4350,
+    PAKID_CORE_DEVICELIST_REMOVE = 0x444D,
+    PAKID_PRN_CACHE_DATA = 0x5043,
+    PAKID_CORE_USER_LOGGEDON = 0x554C,
+    PAKID_PRN_USING_XPS = 0x5543,
+}
+
+impl std::convert::TryFrom<u16> for PacketId {
+    type Error = PduError;
+
+    fn try_from(value: u16) -> Result<Self, Self::Error> {
+        match value {
+            0x496E => Ok(PacketId::PAKID_CORE_SERVER_ANNOUNCE),
+            0x4343 => Ok(PacketId::PAKID_CORE_CLIENTID_CONFIRM),
+            0x434E => Ok(PacketId::PAKID_CORE_CLIENT_NAME),
+            0x4441 => Ok(PacketId::PAKID_CORE_DEVICELIST_ANNOUNCE),
+            0x6472 => Ok(PacketId::PAKID_CORE_DEVICE_REPLY),
+            0x4952 => Ok(PacketId::PAKID_CORE_DEVICE_IOREQUEST),
+            0x4943 => Ok(PacketId::PAKID_CORE_DEVICE_IOCOMPLETION),
+            0x5350 => Ok(PacketId::PAKID_CORE_SERVER_CAPABILITY),
+            0x4350 => Ok(PacketId::PAKID_CORE_CLIENT_CAPABILITY),
+            0x444D => Ok(PacketId::PAKID_CORE_DEVICELIST_REMOVE),
+            0x5043 => Ok(PacketId::PAKID_PRN_CACHE_DATA),
+            0x554C => Ok(PacketId::PAKID_CORE_USER_LOGGEDON),
+            0x5543 => Ok(PacketId::PAKID_PRN_USING_XPS),
+            _ => Err(invalid_message_err!("try_from", "PacketId", "invalid value")),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum VersionAndIdPDUKind {
+    /// [2.2.2.2 Server Announce Request (DR_CORE_SERVER_ANNOUNCE_REQ)](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/046047aa-62d8-49f9-bf16-7fe41880aaf4)
+    ServerAnnounceRequest,
+    /// [2.2.2.3 Client Announce Reply (DR_CORE_CLIENT_ANNOUNCE_RSP)](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/d6fe6d1b-c145-4a6f-99aa-4fe3cdcea398)
+    ClientAnnounceReply,
+}
+
+#[derive(Debug)]
+/// VersionAndIdPDU is a fixed size structure representing multiple PDUs.
+/// The kind field is used to determine the actual PDU type.
```

Comment:
Docstring is typically above the annotations such as `#[derive]`.
Also, [the summary is separated from the rest by a blank line](https://doc.rust-lang.org/rustdoc/how-to-write-documentation.html#documenting-components).
```suggestion
/// VersionAndIdPDU is a fixed size structure representing multiple PDUs.
///
/// The kind field is used to determine the actual PDU type.
#[derive(Debug)]
```